### PR TITLE
Adding a parse_chat step to core agent loop

### PR DESCRIPTION
**Pull request by anuragprat1k**
# Description

Issue
* In the spirit of our recent [refactor push](https://github.com/facebookresearch/droidlet/issues/337), we want the perceptual modules to be stateless. But the current [perceive() ](https://github.com/facebookresearch/droidlet/blob/main/agents/loco_mc_agent.py#L281) implementation in loco_mc_agent does not allow for a generic way for agents to pass inputs to their perceptual modules. For reference, it calls the perceptual modules as  

```
for v in self.perception_modules.values():
            v.perceive(force=force)
``` 
* Moreover, [loco_mc_agent.perceive()](https://github.com/facebookresearch/droidlet/blob/main/agents/loco_mc_agent.py#L281) also processes the incoming chat. This entanglement does not allow for a generic way to fetch inputs that `self.perception_modules` might need.

Fix
* I pulled out the chat perceive into a separate function `parse_chat`. So the core event loop has an additional step now which is separate from the rest of perceive.
* This allows downstream agents to implement their custom perceive functions to fetch states that perception might need, or use the default one if they don't need to. The locobot agent implements a custom perceive already to fetch states for its perceptual modules. [See here](https://github.com/facebookresearch/droidlet/blob/main/agents/locobot/locobot_agent.py#L220).

Fixes # (issue)

This also fixes https://github.com/facebookresearch/droidlet/issues/577 

## Type of change

Please check the options that are relevant.

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [x] Refactor
- [ ] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [ ] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [x] I want a deep design review.

# Testing

Tested that commands now work on the locobot agent 
Screen Shot 2021-07-28 at 7 11 04 PM

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.

### CODE REVIEW

File: agents/core.py [hunk 1]
```diff
@@ -31,6 +32,12 @@ def step(self):
         self.task_step()
         self.count += 1
 
+    def parse_chat(self):
```
Comment (aszlam) [inline agents/core.py:35]: why do we need this here?  sorry I don't understand, these are supposed to go in .perceive()

File: agents/loco_mc_agent.py [hunk 2]
```diff
@@ -278,7 +278,7 @@ def get_time(self):
         # n hundreth of seconds since agent init
         return self.memory.get_time()
 
-    def perceive(self, force=False):
+    def parse_chat(self):
```
Comment (aszlam) [inline agents/loco_mc_agent.py:281]: same with this... these are supposed to go in the .perceive()

**Comment by aszlam** [comment]
ok, at a high level, I strongly disagree with the approach in the PR- semantic parsing is supposed to go in the .perceive().  if the problem is that some state is needed, store whatever state is needed in memory.   to me this looks like just pulling the bits you don't like out of perceive and putting them in some other method, rather than fixing the issue...

**Comment by aszlam** [comment]
look here:
https://github.com/facebookresearch/droidlet/blob/0ee6fa533318802538395203d53b295e2954382e/agents/loco_mc_agent.py#L310


in the craftassist agent were are doing this by calling super.perceive() here
https://github.com/facebookresearch/droidlet/blob/main/agents/craftassist/craftassist_agent.py#L234

**Comment by aszlam** [comment]
In a few iterations, all of this will be properly handled by agent in the sense that all memory writes and reads from the parser will be mediated by the agent- for now, I think its best to keep this on locobot/robot how it is in craftassist agent

**Comment by anuragprat1k** [comment]
Fixing it in https://github.com/facebookresearch/droidlet/pull/614 instead.

**Pull request closed without merging**